### ISSUE TITLE
LLVM assertion opening debuginfo

### ISSUE THREAD

**Issue by KristofferC** [JuliaLang/julia#42607]
Here are the steps to reproduce without PackageCompiler:

```julia
using Pkg

####################
# Create app files #
####################
Pkg.generate("App") # create a package with some context

app_content = """
module App

# Just print out some interesting stuff
Base.@ccallable function julia_main()::Cint
    println("Hello from app")
    @show ARGS
    @show Base.PROGRAM_FILE
    @show DEPOT_PATH
    @show LOAD_PATH
    @show pwd()
    @show Base.active_project()
    @show Sys.BINDIR

    @show Base.JLOptions().opt_level
    @show Base.JLOptions().nthreads
    @show Base.JLOptions().check_bounds

    return 0
end

end # module
"""

write("App/src/App.jl", app_content)

##########################
# Create "driver" C code #
##########################

write("main.c", raw"""
#include <string.h>
#include <stdint.h>
#include <stdio.h>

#include "uv.h"
#include "julia.h"

JULIA_DEFINE_FAST_TLS

int julia_main();

// main function (windows UTF16 -> UTF8 argument conversion code copied from julia's ui/repl.c)
int main(int argc, char *argv[])
{
    jl_options.image_file = "sys.so";
    julia_init(JL_IMAGE_JULIA_HOME);

    julia_main();

    // Cleanup and gracefully exit
    jl_atexit_hook(0);
    return 0;
}
""")

##########################
# Create custom sysimage #
##########################
# Create an object file for a new sysimage with the `App` in it:
sysimg = unsafe_string(Base.JLOptions().image_file)
project = joinpath(@__DIR__, "App")
    julia_path = joinpath(Sys.BINDIR, Base.julia_exename())

code = """
    Base.reinit_stdio()
    copy!(LOAD_PATH, [$(repr(project))])
    import App
"""
cmd_o = `$(julia_path) --startup-file=no --cpu-target='native' --sysimage=$sysimg --output-o=app_sysimg.o -e $code`
run(cmd_o)

# Create the shared library
using Libdl
JULIA_DIR = dirname(dirname(abspath(Libdl.dlpath("libjulia"))))

cmd_o = `gcc -shared -L$JULIA_DIR/lib -L $JULIA_DIR/lib/julia -o sys.so -Wl,--whole-archive app_sysimg.o -Wl,--no-whole-archive -ljulia-internal -ljulia -fPIC`
run(cmd_o)

#####################
# Create executable #
#####################

rpaths = "-Wl,-rpath,$JULIA_DIR/lib:$JULIA_DIR/lib/julia"
cmd_exe = `gcc -o MyApp main.c ./sys.so -O2 $rpaths -std=gnu99 -I$JULIA_DIR/include/julia -fPIC -L$JULIA_DIR/lib -L$JULIA_DIR/lib/julia -Wl,--export-dynamic -Wl,-rpath-link,$JULIA_DIR/lib -ljulia -ljulia-internal`
run(cmd_exe)

######################
# Run the executable #
######################

run(`./MyApp`)
```

This errors with the LLVM assertion

```
MyApp: /home/kc/julia/usr/include/llvm/Support/ErrorOr.h:237: llvm::ErrorOr<T>::storage_type* llvm::ErrorOr<T>::getStorage() [with T = std::unique_ptr<llvm::MemoryBuffer>; llvm::ErrorOr<T>::storage_type = std::unique_ptr<llvm::MemoryBuffer>]: Assertion `!HasError && "Cannot get value when an error exists!"' failed.
```

Full output is:

```
Hello from app
ARGS = String[]
Base.PROGRAM_FILE = ""
DEPOT_PATH = ["/home/kc/.julia", "/home/kc/JuliaPkgs/PackageCompilerRepro.jl/local/share/julia", "/home/kc/JuliaPkgs/PackageCompilerRepro.jl/share/julia"]
LOAD_PATH = ["@", "@v#.#", "@stdlib"]
pwd() = "/home/kc/JuliaPkgs/PackageCompilerRepro.jl/Repro2"
Base.active_project() = "/home/kc/.julia/environments/v1.8/Project.toml"
Sys.BINDIR = "/home/kc/JuliaPkgs/PackageCompilerRepro.jl/Repro2"

signal (11): Segmentation fault
in expression starting at none:0
string at ./intfuncs.jl:780 [inlined]
show at ./show.jl:1054
unknown function (ip: 0x7f0ca4e27279)
#sprint#432 at ./strings/io.jl:114
unknown function (ip: 0x7f0ca4e271c4)
sprint##kw at ./strings/io.jl:108 [inlined]
#repr#435 at ./strings/io.jl:282 [inlined]
repr at ./strings/io.jl:282 [inlined]
julia_main at /home/kc/JuliaPkgs/PackageCompilerRepro.jl/Repro2/App/src/App.jl:14
jl_sysimg_fvars_base at ./sys.so (unknown line)
julia_main at ./sys.so (unknown line)
main at ./MyApp (unknown line)
MyApp: /home/kc/julia/usr/include/llvm/Support/ErrorOr.h:237: llvm::ErrorOr<T>::storage_type* llvm::ErrorOr<T>::getStorage() [with T = std::unique_ptr<llvm::MemoryBuffer>; llvm::ErrorOr<T>::storage_type = std::unique_ptr<llvm::MemoryBuffer>]: Assertion `!HasError && "Cannot get value when an error exists!"' failed.

signal (6): Aborted
in expression starting at none:0
gsignal at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
abort at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
unknown function (ip: 0x7f0d3a245748)
__assert_fail at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
getStorage at /home/kc/julia/usr/include/llvm/Support/ErrorOr.h:237 [inlined]
get at /home/kc/julia/usr/include/llvm/Support/ErrorOr.h:150 [inlined]
openDebugInfo at /home/kc/julia/src/debuginfo.cpp:691
find_object_file at /home/kc/julia/src/debuginfo.cpp:981
jl_dylib_DI_for_fptr at /home/kc/julia/src/debuginfo.cpp:1129
jl_getDylibFunctionInfo at /home/kc/julia/src/debuginfo.cpp:1164 [inlined]
jl_getFunctionInfo_impl at /home/kc/julia/src/debuginfo.cpp:1232
jl_print_native_codeloc at /home/kc/julia/src/stackwalk.c:636
jl_critical_error at /home/kc/julia/src/signal-handling.c:286
unknown function (ip: 0x7f0d3a26003f)
string at ./intfuncs.jl:780 [inlined]
show at ./show.jl:1054
unknown function (ip: 0x7f0ca4e27279)
#sprint#432 at ./strings/io.jl:114
unknown function (ip: 0x7f0ca4e271c4)
sprint##kw at ./strings/io.jl:108 [inlined]
#repr#435 at ./strings/io.jl:282 [inlined]
repr at ./strings/io.jl:282 [inlined]
julia_main at /home/kc/JuliaPkgs/PackageCompilerRepro.jl/Repro2/App/src/App.jl:14
jl_sysimg_fvars_base at ./sys.so (unknown line)
julia_main at ./sys.so (unknown line)
main at ./MyApp (unknown line)
__libc_start_main at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
_start at ./MyApp (unknown line)
Allocations: 2730 (Pool: 2719; Big: 11); GC: 0
ERROR: failed process: Process(`./MyApp`, ProcessSignaled(6)) [0]
```

**Comment by KristofferC** [comment]
I'm looking at the command line that gets executed when building Julia locally:

```
 g++ -m64  -shared -fPIC -L/home/kristofferc/julia/usr/lib/julia -L/home/kristofferc/julia/usr/lib -L/home/kristofferc/julia/usr/lib -o /home/kristofferc/julia/usr/lib/julia/sys.so -Wl,--whole-archive /home/kristofferc/julia/usr/lib/julia/sys-o.a -Wl,--no-whole-archive -ljulia-internal -ljulia $([ Linux = WINNT ] && echo '' -lssp)
```

So there, the sysimage doesn't seem to link against `pthread`. So what is the difference?

**Comment by vtjnash** [comment]
That is fixed by 2268ee47e25503a4d8b15c1cd9044fac2f19bec3.

Has anyone filed this bug with glibc? It is also possible they already fixed this, since they recently redesigned this entirely and replaced the prior unreliable weak implementation with a dynamic `__libc_single_threaded` flag which perhaps already fixed this (released Aug 2, 2021).
